src/librustc/middle/const_eval.rs

```diff
@@ -1007,14 +1022,76 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
           } else {
               ty_hint
           };
-          try!(eval_const_expr_partial(tcx, const_expr, item_hint))
+          try!(eval_const_expr_partial(tcx, const_expr, item_hint, fn_args))
       }
+      hir::ExprCall(ref callee, ref args) => {
+          let sub_ty_hint = if let ExprTypeChecked = ty_hint {
+              ExprTypeChecked
+          } else {
+              UncheckedExprNoHint // we cannot reason about UncheckedExprHint here
+          };
+          let (
+              decl,
+              unsafety,
+              abi,
+              block,
+              constness,
+          ) = match try!(eval_const_expr_partial(tcx, callee, sub_ty_hint, fn_args)) {
+              Function(did) => if did.is_local() {
+                  match tcx.map.find(did.index.as_u32()) {
+                      Some(ast_map::NodeItem(it)) => match it.node {
+                          hir::ItemFn(
+                              ref decl,
+                              unsafety,
+                              constness,
+                              abi,
+                              _, // ducktype generics? types are funky in const_eval
+                              ref block,
+                          ) => (decl, unsafety, abi, block, constness),
+                          _ => signal!(e, NonConstPath),
+                      },
+                      _ => signal!(e, NonConstPath),
+                  }
+              } else {
+                  signal!(e, NonConstPath)
+              },
+              _ => signal!(e, NonConstPath),
+          };
+          if let ExprTypeChecked = ty_hint {
+              // no need to check for constness... either check_const
+              // already forbids this or we const eval over whatever
+              // we want
+          } else {
+              // we don't know much about the function, so we force it to be a const fn
+              // so compilation will fail later in case the const fn's body is not const
+              assert_eq!(constness, hir::Constness::Const)
```

Comment:
Great, thanks

On Wed, 4 Nov 2015 at 20:46 eefriedman <notifications@github.com> wrote:

> In src/librustc/middle/const_eval.rs
> <https://github.com/rust-lang/rust/pull/26848#discussion_r43936838>:
>
> > +                      },
> > +                      _ => signal!(e, NonConstPath),
> > +                  }
> > +              } else {
> > +                  signal!(e, NonConstPath)
> > +              },
> > +              _ => signal!(e, NonConstPath),
> > +          };
> > +          if let ExprTypeChecked = ty_hint {
> > +              // no need to check for constness... either check_const
> > +              // already forbids this or we const eval over whatever
> > +              // we want
> > +          } else {
> > +              // we don't know much about the function, so we force it to be a const fn
> > +              // so compilation will fail later in case the const fn's body is not const
> > +              assert_eq!(constness, hir::Constness::Const)
>
> Filed #29587 <https://github.com/rust-lang/rust/issues/29587>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/26848/files#r43936838>.
>